### PR TITLE
Fix FP reported in #467

### LINKED ISSUE #467: `M9-3-3`: Do not flag uninstantiated templates
### Affected rules
 - `M9-3-3`

### Description

Uninstantiated templates may not have resolved all the calls, which prevents this analysis from determining whether non-const functions are further called.

### Example

```cpp

template<typename T>
class Wrapper {
public:
  T& Top() { return elements.back(); } // Likely NON_COMPLIANT, but cannot be determined until instantiation.
private:
  vector<T> elements;
};

```

### PR DESCRIPTION
**Pull request by rvermeulen**
## Description

Fix #467 

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - M9-3-3

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: cpp/autosar/test/rules/M9-3-3/test.cpp [hunk 1]
```diff
@@ -161,3 +161,29 @@ class Z22 : Z1 {
...
+private:
+  T data[128];
+  unsigned int size;
+};
+
+template <class T, template <class...> class U> class Stack {
+public:
+  T &Top() {
+    return this->data.back();
+  } // Likely NON_COMPLIANT, but cannot be determined until instantiation.
```
Comment (knewbury01) [inline cpp/autosar/test/rules/M9-3-3/test.cpp:178]: lets change this message to best describe the expected behaviour (align with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#style-guide)), 

in this case with the instantiation, it is a COMPLIANT example due to (after instantiation) being able to see that it calls a non const using member variable.

though this location represents 2 member functions, and one (the instantiated one) is compliant due to the logic of the rule
and the non instantiated one is compliant due to not having enough info to know if it obeys the rule - 

so maybe... label COMPLIANT - instantiated compliant truly and uninstantiated compliant via exception
or whatever seems most clear to describe the expected behaviour

**Review by knewbury01 — CHANGES_REQUESTED** [review]
minor change requested to make the testcase slightly more clear 

(since the representations in the db are not the most clear around templates - ie this only shows up once the template is instantiated, but its the uninstantiated version of the function that was falsley triggering it 😅 )

File: cpp/autosar/test/rules/M9-3-3/test.cpp (hunk 1 above)
Comment (rvermeulen) [reply cpp/autosar/test/rules/M9-3-3/test.cpp:178]: I would argue that the style guide is inconclusive about cases that are neither compliant nor non-compliant. Perhaps we should call it `COMPLIANT[FALSE_NEGATIVE|TRUE_NEGATIVE]` to highlight a nonstandard exception with the explanation that the template dictates the interface and not its instantiations?

Comment (knewbury01) [reply cpp/autosar/test/rules/M9-3-3/test.cpp:178]: I would be happy with either a `COMPLIANT - exclude consideration of functions in uninstantiated templates` or `COMPLIANT[FALSE_NEGATIVE|TRUE_NEGATIVE] - insert explanation` with the explanation you mention

its true that the latter doesnt completely follow the style guide, so I am leaning towards the former but would find either to communicate what is going on well enough (at least imo labelling it both fn/tn indicates that we are purposely ignoring it)

File: cpp/autosar/test/rules/M9-3-3/test.cpp [hunk 2]
```diff
@@ -162,6 +162,35 @@ class Z22 : Z1 {
...
+  IntVectorStack s;
+
+  int i = s.Top();
+}
+
 class Z3 {
   void f(int) = delete; // COMPLIANT
-};
\ No newline at end of file
+};
```
Comment (knewbury01) [inline cpp/autosar/test/rules/M9-3-3/test.cpp:195]: needs a reformat, then it will be all good

**Review by knewbury01 — CHANGES_REQUESTED** [review]

**Review by knewbury01 — APPROVED** [review]
nice work!

**Pull request merged**